design/one-pager-managed-resource-api-design.md

```diff
@@ -330,6 +403,69 @@ type GlobalAddressParameters struct {
 }
 ```
 
+#### Optional Embedded Structs with One Field
+
+Some provider APIs include an embedded struct that is optional and contains only
+a single field. If this field is configurable, the embedded struct should be
+part of `spec.forProvider`. However, if the struct itself is optional, the
+single field it contains should always be **required**, *even if the provider
+marks it as optional*. For example:
+
+```go
+// This is the provider's representation of the API object
+type ProviderAPIObject struct {
+  // Configurable field in top-level object
+  FieldOne *string `json:"fieldOne,omitempty"`
+  
+  // Embedded struct in top-level object
+  EmbeddedStructOne *EmbeddedStruct `json:"embeddedStructOne,omitempty"`
+}
+
+type EmbeddedStruct struct {
+  // This field is configurable so it should be in spec.forProvider
+  SomeConfigurableField *string `json:"someConfigurableField,omitempty"`
+}
+```
+
+The corresponding Crossplane representation of the object's `Spec` should look
+as follows:
+
+```go
+// This is the Crossplane representation of the API object spec
+type CrossplaneAPIObjectSpec struct {
+  // Configurable field in top-level object
+  // +optional
+  FieldOne *string `json:"fieldOne,omitempty"`
+  
+  // Embedded struct in top-level object
+  // +optional
+  EmbeddedStructOne *EmbeddedStructSpec `json:"embeddedStructOne,omitempty"`
+}
+
+// Only the configurable fields in EmbeddedStruct
+type EmbeddedStructSpec struct {
+  // This field is configurable so it should be in spec.forProvider
+  // It is required because its parent is optional and it is the only field of its parent
+  SomeConfigurableField string `json:"someConfigurableField"`
+}
+```
+
+The reasoning for designing in this manner is to avoid the presence of
+meaningless entries in `YAML` configuration. If we *did not* do so, the
+following configuration would be valid:
+
+```yaml
+...
+spec:
+  providerRef:
+    name: provider-gcp
+  reclaimPolicy: Delete
+  writeConnectionSecretToRef: {}
+  forProvider:
+    embeddedStructOne: # valid if we do not require embedded field
+    fieldOne: my-cool-field
+```
+
```

Comment:
Do we need to consider the case in which additional optional fields are added over time to the cloud provider struct?

I could imagine a case in which:

1. We mark field A, which is optional to the cloud provider, required.
1. The cloud provider adds field B, which is also optional.
1. Our users now cannot specify field B without also specifying field A.

My inclination would be to avoid overriding the cloud provider here, even if it does mean certain nonsense configurations such as the above become valid.

Reply:
Hmm, in that case our API would have to be updated to include that additional field, and at that point we would no longer have the situation described here. In addition, I think we have already implemented in this manner as `omitempty` is removed in our API where it is present in GCP: https://github.com/crossplaneio/stack-gcp/blob/f7522e08bccbf856cb1cd3ad713efe597f45785b/apis/database/v1beta1/cloudsql_instance_types.go#L569

Changing this *would* require an update to the API.

Reply:
I think this update would require us to make a required field optional, which should be a backward compatible change. I think I was getting mixed up and thinking we'd have to make an optional field required and thus break API compatibility. This seems fine to me, though TBH I'm not super concerned about the problem it's solving.